File: app/src/main/java/com/juniperphoton/myerlistandroid/presenter/MainPresenter.java
package com.juniperphoton.myerlistandroid.presenter;

import android.util.Log;

import com.google.gson.Gson;
import com.juniperphoton.myerlistandroid.App;
import com.juniperphoton.myerlistandroid.R;
import com.juniperphoton.myerlistandroid.api.APIException;
import com.juniperphoton.myerlistandroid.api.CloudService;
import com.juniperphoton.myerlistandroid.api.response.AddToDoResponse;
import com.juniperphoton.myerlistandroid.api.response.CateResponse;
import com.juniperphoton.myerlistandroid.api.response.CommonResponse;
import com.juniperphoton.myerlistandroid.api.response.GetOrderResponse;
import com.juniperphoton.myerlistandroid.api.response.GetToDosResponse;
import com.juniperphoton.myerlistandroid.model.CategoryRespInformation;
import com.juniperphoton.myerlistandroid.model.OrderedCateList;
import com.juniperphoton.myerlistandroid.model.ToDo;
import com.juniperphoton.myerlistandroid.model.ToDoCategory;
import com.juniperphoton.myerlistandroid.realm.RealmUtils;
import com.juniperphoton.myerlistandroid.util.AppConfig;
import com.juniperphoton.myerlistandroid.util.ToastService;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmResults;
import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

public class MainPresenter implements MainContract.Presenter {

    private static final String TAG = "MainPresenter";

    private MainContract.View mView;

    public MainPresenter(MainContract.View mainView) {
        mView = mainView;
    }

    @Override
    public void getCates() {
        CloudService.getInstance().getCategories()
                .subscribeOn(Schedulers.io())
                .map(new Func1<CateResponse, Object>() {
                    @Override
                    public Object call(CateResponse cateResponse) {
                        parseCategories(cateResponse.getRawString());
                        return null;
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<Object>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                    }

                    @Override
                    public void onNext(Object cateResponse) {
                        mView.displayCategories();
                        getToDos();
                    }
                });
    }

    @Override
    public void getToDos() {
        CloudService.getInstance().getToDos()
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .flatMap(new Func1<GetToDosResponse, Observable<GetOrderResponse>>() {
                    @Override
                    public Observable<GetOrderResponse> call(final GetToDosResponse toDoResponse) {
                        if (toDoResponse.getToDos() != null) {
                            mergeToDos(toDoResponse.getToDos());
                            return CloudService.getInstance().getOrders();
                        } else {
                            return Observable.error(new APIException(toDoResponse.getFriendErrorMessage()));
                        }
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<GetOrderResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(GetOrderResponse getOrderResponse) {
                        Log.d(TAG, "got order");
                        orderToDos(getOrderResponse);
                    }
                });
    }

    private void mergeToDos(List<ToDo> toDos) {
        Realm realm = RealmUtils.getMainInstance();
        realm.beginTransaction();
        RealmResults<ToDo> shouldDeleted = realm.where(ToDo.class)
                .notEqualTo(ToDo.DELETED_KEY, Boolean.TRUE).findAll();
        for (ToDo toDo : shouldDeleted) {
            toDo.deleteFromRealm();
        }
        for (ToDo toDo : toDos) {
            realm.copyToRealmOrUpdate(toDo);
        }
        realm.commitTransaction();

        Log.d(TAG, "got to-dos");
    }

    @Override
    public void updateOrders(String order) {
        CloudService.getInstance().setOrders(order).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Action1<CommonResponse>() {
                    @Override
                    public void call(CommonResponse commonResponse) {
                        Log.d(TAG, "updateOrders");
                    }
                });
    }

    @Override
    public void updateIsDone(ToDo toDo) {
        CloudService.getInstance().setIsDone(toDo.getId(), toDo.getIsdone()).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Action1<CommonResponse>() {
                    @Override
                    public void call(CommonResponse commonResponse) {
                        Log.d(TAG, "updateIsDone");
                    }
                });
    }

    @Override
    public void modifyToDo(final String cate, final String content, final String id) {
        final String dateStr = getDateStr();
        Realm realm = RealmUtils.getMainInstance();
        realm.beginTransaction();
        ToDo toDo = realm.where(ToDo.class).equalTo("id", id).findFirst();
        if (toDo != null) {
            toDo.setContent(content);
            toDo.setCate(cate);
            toDo.setTime(dateStr);
        }
        realm.commitTransaction();
        CloudService.getInstance().updateToDo(id, dateStr, content, cate)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<CommonResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(CommonResponse commonResponse) {
                        if (commonResponse.getOK()) {
                            ToastService.sendShortToast(App.getInstance().getString(R.string.modified_hint));
                        } else {
                            ToastService.sendShortToast(commonResponse.getFriendErrorMessage());
                            Log.e(TAG, "Modify failed:" + commonResponse.getFriendErrorMessage());
                        }
                    }
                });
    }

    private String getDateStr() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        Date date = new Date();
        return sdf.format(date);
    }

    @Override
    public void addToDo(String cate, String content) {
        String dateStr = getDateStr();
        CloudService.getInstance().addToDo(dateStr, content, ToDo.IS_NOT_DONE, cate)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<AddToDoResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                        ToastService.sendShortToast(e.getMessage());
                    }

                    @Override
                    public void onNext(AddToDoResponse addToDoResponse) {
                        ToDo toDo = addToDoResponse.getToDo();
                        if (toDo != null) {
                            Realm realm = RealmUtils.getMainInstance();
                            realm.beginTransaction();

                            ToDo managedToDo = realm.copyToRealm(toDo);

                            if (AppConfig.addToBottom()) {
                                int pos = realm.where(ToDo.class).findAll().max(ToDo.POSITION_KEY).intValue();
                                managedToDo.setPosition(++pos);
                            } else {
                                int pos = realm.where(ToDo.class).findAll().min(ToDo.POSITION_KEY).intValue();
                                managedToDo.setPosition(--pos);
                            }
                            realm.commitTransaction();

                            mView.displayToDos();
                            mView.uploadOrders();
                        }
                    }
                });
    }

    @Override
    public void deleteToDo(ToDo toDo) {
        String id = toDo.getId();

        Realm realm = RealmUtils.getMainInstance();
        realm.beginTransaction();
        toDo.setDeleted(true);
        realm.commitTransaction();

        mView.notifyToDoDeleted(toDo.getPosition());

        CloudService.getInstance().deleteToDo(id).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Action1<CommonResponse>() {
                    @Override
                    public void call(CommonResponse commonResponse) {
                        Log.d(TAG, "deleteToDo");
                    }
                });
    }

    @Override
    public void clearDeletedList() {
        Realm realm = RealmUtils.getMainInstance();
        realm.beginTransaction();
        RealmResults<ToDo> deletedList = realm.where(ToDo.class).equalTo(ToDo.DELETED_KEY, true)
                .findAll();
        for (ToDo todo : deletedList) {
            todo.deleteFromRealm();
        }
        realm.commitTransaction();

        mView.displayToDos();
    }

    @Override
    public void recoverToDo(ToDo toDo) {
        addToDo(toDo.getCate(), toDo.getContent());
        Realm realm = RealmUtils.getMainInstance();
        realm.beginTransaction();
        toDo.deleteFromRealm();
        realm.commitTransaction();
        mView.displayToDos();
    }

    private void orderToDos(GetOrderResponse getOrderResponse) {
        String order = getOrderResponse.getOrder();
        if (order != null) {
            final String[] orders = order.split(",");

            Realm realm = RealmUtils.getMainInstance();
            realm.beginTransaction();

            RealmResults<ToDo> localToDos = realm.where(ToDo.class).findAll();
            for (int i = 0; i < orders.length; i++) {
                ToDo toDo = localToDos.where().equalTo(ToDo.ID_KEY, orders[i]).findFirst();
                if (toDo != null) {
                    toDo.setPosition(i);
                }
            }

            realm.commitTransaction();
            ToastService.sendShortToast(App.getInstance().getString(R.string.fetch_hint));
            mView.displayToDos();
        } else {
            Log.d(TAG, "Can't get order.");
        }
    }

    private void parseCategories(String resp) {
        Gson gson = new Gson();
        final CategoryRespInformation information = gson.fromJson(resp, CategoryRespInformation.class);
        if (information.isModified()) {
            RealmUtils.getMainInstance().executeTransaction(new Realm.Transaction() {
                @Override
                public void execute(Realm realm) {
                    OrderedCateList orderedCateList = new OrderedCateList();
                    RealmList<ToDoCategory> toDoCategories = orderedCateList.getCates();
                    for (ToDoCategory cate : information.getCates()) {
                        toDoCategories.add(cate);
                    }
                    RealmUtils.getMainInstance().copyToRealmOrUpdate(orderedCateList);
                }
            });
        } else {
            //// TODO: 12/24/2016  handle not modified situation
        }
    }

    @Override
    public void start() {

    }

    @Override
    public void stop() {

    }
}

File: app/src/main/java/com/juniperphoton/myerlist/presenter/MainPresenter.kt
package com.juniperphoton.myerlist.presenter

import android.util.Log
import com.google.gson.Gson
import com.juniperphoton.myerlist.App
import com.juniperphoton.myerlist.R
import com.juniperphoton.myerlist.api.APIException
import com.juniperphoton.myerlist.api.CloudService
import com.juniperphoton.myerlist.api.response.AddToDoResponse
import com.juniperphoton.myerlist.api.response.CommonResponse
import com.juniperphoton.myerlist.api.response.GetOrderResponse
import com.juniperphoton.myerlist.event.ReCreateEvent
import com.juniperphoton.myerlist.event.RefreshToDoEvent
import com.juniperphoton.myerlist.extension.getResString
import com.juniperphoton.myerlist.model.CategoryRespInformation
import com.juniperphoton.myerlist.model.ToDo
import com.juniperphoton.myerlist.model.ToDoCategory
import com.juniperphoton.myerlist.util.AppConfig
import com.juniperphoton.myerlist.util.LocalSettingUtil
import com.juniperphoton.myerlist.util.Params
import com.juniperphoton.myerlist.util.ToastService
import io.realm.Realm
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import rx.Observable
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.text.SimpleDateFormat
import java.util.*

class MainPresenter(private val view: MainContract.View) : MainContract.Presenter {
    companion object {
        private val TAG = "MainPresenter"

        private val DEFAULT = "{ \\\"modified\\\":true, \\\"cates\\\":[{\\\"name\\\":\\\"Work\\\",\\\"color\\\":\\\"#FF436998\\\",\\\"id\\\":1},{\\\"name\\\":\\\"Life\\\",\\\"color\\\":\\\"#FFFFB542\\\",\\\"id\\\":2},{\\\"name\\\":\\\"Family\\\",\\\"color\\\":\\\"#FFFF395F\\\",\\\"id\\\":3},{\\\"name\\\":\\\"Entertainment\\\",\\\"color\\\":\\\"#FF55C1C1\\\",\\\"id\\\":4}]}"

        const val FILTER_ALL = 0
        const val FILTER_UNDONE = 1
        const val FILTER_DONE = 2
    }

    override var filterOption: Int = LocalSettingUtil.getInt(App.instance!!, Params.KEY_FILTER_OPTION, FILTER_ALL)
        set(value) {
            if (field != value) {
                field = value
                view.refreshToDoList(value)
                view.updateFilterIcon(value)

                LocalSettingUtil.putInt(App.instance!!, Params.KEY_FILTER_OPTION, value)
            }
        }

    override fun onCategorySelected(item: Int) {
        view.refreshToDoList(filterOption)
    }

    override fun getCategories() {
        view.toggleRefreshing(true)
        CloudService.getCategories()
                .subscribeOn(Schedulers.io())
                .map { cateResponse ->
                    parseCategories(cateResponse?.rawString)
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<Unit>() {
                    override fun onCompleted() {
                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                        view.toggleRefreshing(false)
                        view.refreshToDoList(filterOption)
                    }

                    override fun onNext(cateResponse: Unit) {
                        view.refreshCategoryList()
                        getToDos()
                    }
                })
    }

    override fun getToDos() {
        view.toggleRefreshing(true)
        CloudService.getToDos()
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .flatMap({ toDoResponse ->
                    if (toDoResponse.toDos != null) {
                        mergeToDos(toDoResponse.toDos)
                        CloudService.getOrders()
                    } else {
                        Observable.error<GetOrderResponse>(APIException(toDoResponse.friendErrorMessage))
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<GetOrderResponse>() {
                    override fun onCompleted() {
                        view.toggleRefreshing(false)
                        view.refreshToDoList(filterOption)
                    }

                    override fun onError(e: Throwable) {
                        view.toggleRefreshing(false)
                        view.refreshToDoList(filterOption)
                    }

                    override fun onNext(getOrderResponse: GetOrderResponse) {
                        Log.d(TAG, "got order")
                        orderToDos(getOrderResponse)
                    }
                })
    }

    private fun mergeToDos(toDos: List<ToDo>) {
        Realm.getDefaultInstance().executeTransaction { realm ->
            val localWithoutDeleteFlag = realm.where(ToDo::class.java)
                    .notEqualTo(ToDo.KEY_DELETED, java.lang.Boolean.TRUE).findAll()
            localWithoutDeleteFlag.forEach {
                it.deleteFromRealm()
            }
            toDos.forEach {
                realm.copyToRealmOrUpdate(it)
            }
        }
        Log.d(TAG, "mergeToDos")
    }

    override fun updateOrders(order: String) {
        CloudService.setOrders(order).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<CommonResponse>() {
                    override fun onError(e: Throwable?) {
                        e?.printStackTrace()
                    }

                    override fun onCompleted() {
                    }

                    override fun onNext(t: CommonResponse?) {
                        Log.d(TAG, "updateOrders")
                    }
                })
    }

    override fun updateIsDone(id: String, oldValue: String) {
        val newValue = if (oldValue == ToDo.VALUE_DONE) ToDo.VALUE_UNDONE else ToDo.VALUE_DONE

        val realm = Realm.getDefaultInstance()
        realm.beginTransaction()
        val managedToDo = realm.where(ToDo::class.java).equalTo(ToDo.KEY_ID, id).findFirst()
        managedToDo!!.isdone = newValue
        realm.commitTransaction()

        view.notifyDataSetChanged()

        CloudService.setIsDone(id, newValue).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<CommonResponse>() {
                    override fun onError(e: Throwable?) {
                        e?.printStackTrace()
                    }

                    override fun onCompleted() {
                    }

                    override fun onNext(t: CommonResponse?) {
                    }
                })
    }

    override fun modifyToDo(id: String, cate: String, content: String) {
        val dateStr = dateStr
        Realm.getDefaultInstance().executeTransaction {
            val toDo = it.where(ToDo::class.java).equalTo(ToDo.KEY_ID, id).findFirst()
            toDo?.let {
                it.content = content
                it.cate = cate
                it.time = dateStr
            }
        }

        view.notifyDataSetChanged()
        view.toggleRefreshing(true)

        CloudService.updateToDo(id, dateStr, content, cate)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<CommonResponse>() {
                    override fun onCompleted() {
                    }

                    override fun onError(e: Throwable) {
                        view.toggleRefreshing(false)
                        e.printStackTrace()
                    }

                    override fun onNext(commonResponse: CommonResponse) {
                        view.toggleRefreshing(false)
                        if (commonResponse.ok) {
                            ToastService.sendShortToast(R.string.modified_hint.getResString())
                        } else {
                            ToastService.sendShortToast(commonResponse.friendErrorMessage)
                            Log.e(TAG, "Modify failed:" + commonResponse.friendErrorMessage)
                        }
                    }
                })
    }

    private val dateStr: String
        get() {
            val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
            val date = Date()
            return sdf.format(date)
        }

    override fun addToDo(cate: String, content: String) {
        val dateStr = dateStr
        view.toggleRefreshing(true)
        CloudService.addToDo(dateStr, content, ToDo.VALUE_UNDONE, cate)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<AddToDoResponse>() {
                    override fun onCompleted() {

                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                        ToastService.sendShortToast(e.message!!)
                        view.toggleRefreshing(false)
                    }

                    override fun onNext(addToDoResponse: AddToDoResponse) {
                        view.toggleRefreshing(false)
                        val toDo = addToDoResponse.toDo
                        if (toDo != null) {
                            Realm.getDefaultInstance().executeTransaction {
                                val managedToDo = it.copyToRealm(toDo)
                                if (AppConfig.addToBottom()) {
                                    var pos = it.where(ToDo::class.java).findAll().max(ToDo.KEY_POSITION).toInt()
                                    managedToDo.position = ++pos
                                } else {
                                    var pos = it.where(ToDo::class.java).findAll().min(ToDo.KEY_POSITION).toInt()
                                    managedToDo.position = --pos
                                }
                            }
                            view.refreshToDoList(filterOption)
                            view.uploadOrders()
                        }
                    }
                })
    }

    override fun deleteToDo(item: ToDo) {
        val id = item.id
        val alreadyDeleted = item.deleted

        Realm.getDefaultInstance().executeTransaction { realm ->
            val todo = realm.where(ToDo::class.java).equalTo(ToDo.KEY_ID, id).findFirst()
            todo?.let {
                if (!alreadyDeleted) {
                    it.deleted = true
                } else {
                    it.deleteFromRealm()
                }
            }
        }

        view.notifyDataSetChanged()

        if (!alreadyDeleted) {
            CloudService.deleteToDo(id!!).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(object : Subscriber<CommonResponse>() {
                        override fun onError(e: Throwable?) {
                            e?.printStackTrace()
                        }

                        override fun onCompleted() {
                        }

                        override fun onNext(t: CommonResponse?) {
                            Log.d(TAG, "deleteToDo")
                        }
                    })
        }
    }

    override fun clearDeletedList() {
        Realm.getDefaultInstance().executeTransaction {
            val deletedList = it.where(ToDo::class.java).equalTo(ToDo.KEY_DELETED, true)
                    .findAll()
            for (todo in deletedList) {
                todo.deleteFromRealm()
            }
        }

        view.refreshToDoList(filterOption)
    }

    override fun recoverToDo(toDo: ToDo) {
        addToDo(toDo.cate!!, toDo.content!!)
        Realm.getDefaultInstance().executeTransaction {
            toDo.deleteFromRealm()
        }
        view.refreshToDoList(filterOption)
    }

    private fun orderToDos(getOrderResponse: GetOrderResponse) {
        val order = getOrderResponse.order
        if (order != null) {
            Log.d(TAG, "todo order$order")

            val orders = order.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            Realm.getDefaultInstance().executeTransaction {
                val localToDos = it.where(ToDo::class.java).findAll()
                orders.indices.forEach {
                    val toDo = localToDos.where().equalTo(ToDo.KEY_ID, orders[it]).findFirst()
                    if (toDo != null) {
                        toDo.position = it
                    }
                }
            }

            ToastService.sendShortToast(App.instance!!.getString(R.string.fetch_hint))
        } else {
            Log.d(TAG, "Can't get order.")
        }
    }

    private fun parseCategories(resp: String?) {
        Log.d(TAG, "cate resp$resp")
        val gson = Gson()
        if (resp != null) {
            val information = gson.fromJson(resp, CategoryRespInformation::class.java)
            saveCategoriesToRealm(information!!.cates!!)
        } else {
            val list = gson.fromJson(DEFAULT, CategoryRespInformation::class.java)
            saveCategoriesToRealm(list!!.cates!!)
        }
    }

    private fun saveCategoriesToRealm(list: List<ToDoCategory>) {
        Realm.getDefaultInstance().executeTransaction { realm ->
            for ((i, cate) in list.withIndex()) {
                cate.position = i
                cate.setSid(LocalSettingUtil.getString(App.instance!!, Params.SID_KEY)!!)
                realm.copyToRealmOrUpdate(cate)
            }
        }
    }

    @Suppress("unused_parameter", "unused")
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    fun receiveEvent(event: ReCreateEvent) {
        view.reCreateView()
        EventBus.getDefault().removeAllStickyEvents()
    }

    @Suppress("unused_parameter", "unused")
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    fun receiveRefreshToDoEvent(event: RefreshToDoEvent) {
        view.refreshToDoList(filterOption)
        EventBus.getDefault().removeAllStickyEvents()
    }

    override fun start() {
        EventBus.getDefault().register(this)
        view.updateFilterIcon(filterOption)
    }

    override fun stop() {
        EventBus.getDefault().unregister(this)
    }
}
